PR: https://github.com/odoo/odoo/pull/234455

From: aeaca097ae39b293bff47458ae8af019585f9224
From: David Monnom
Date: 2026-01-15 11:40:50

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = ["portal.mixin", "pos.bus.mixin", "pos.load.mixin", "mail.thread"]

Total Changes: 107

[REF] point_of_sale, pos_*: ticket generation

*: l10n_ae_pos, l10n_co_pos, l10n_es_edi_tbai_pos,
   l10n_es_edi_verifactu_pos, l10n_es_pos, l10n_fr_pos_cert,
   l10n_gcc_pos, l10n_in_pos, l10n_sa_edi_pos, l10n_sa_pos, pos_loyalty
   pos_restaurant, pos_self_order

This commit introduces a new `point_of_sale.pos_order_receipt` template
that can be used to generate receipts via the QWeb Python & JS renderer.

The main idea was to avoid duplicating the receipt template, as it is
overridden in almost all modules related to the Point of Sale.
Duplicating this template would have been tedious to maintain.

As a result, this new template stored in the Point of Sale data is a
"simple" template in which we cannot use directives that are only
understandable by Python or JS. Here are the directives that can be used
* `t-if`
* `t-elif`
* `t-else`
* `t-foreach`
* `t-att-*`
* `t-attf-*`
* `t-call`
* `t-out`
* `t-set`
* `t-as`
* `t-name`
* `t-value`

This template then takes a pre-built object from a command and retrieves
all the necessary information from it. It is important to remember that
JS and Python must generate the same object.

*--

About evaluations:
JS rendering is more permissive than Python. For example, calling an
undefined index `order['lines']['data']` will be accepted by JS but not
by Python.

We therefore assume that Python will be the reference when modifying
the template.

Complex evaluations in Python/JS are not allowed, as they require overly
complex conversion.

*--

Template conversion:
When downloading the template in JS, these conversions are performed:
* Addition of `t-key` for `t-foreach`
* Replacement of `.get('something')` with `['something']`
* The operator `and` becomes `&amp;&amp;`.
* The `or` operator becomes `||`.
* The `not` operator becomes `!`.
* `True` becomes `true`
* `False` becomes `false`
* `None` becomes `null`

This conversion is simple and matches the template's requirements.
The method used cannot be used with other ASIS backend templates.

closes odoo/odoo#234455

Taskid: 5113480
Related: odoo/enterprise#99109
Signed-off-by: Stéphane Vanmeerhaeghe (stva) <stva@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/pos_order.py
+++ b/addons/point_of_sale/models/pos_order.py
@@ -1,6 +1,7 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 import logging
 import json
+import base64
 from datetime import datetime, UTC
 from itertools import groupby
 from collections import defaultdict
@@ -22,7 +23,7 @@ _logger = logging.getLogger(__name__)
 
 class PosOrder(models.Model):
     _name = 'pos.order'
-    _inherit = ["portal.mixin", "pos.bus.mixin", "pos.load.mixin", "mail.thread"]
+    _inherit = ["portal.mixin", "pos.bus.mixin", "pos.order.receipt", "pos.load.mixin", "mail.thread"]
     _description = "Point of Sale Order"
     _order = "date_order desc, name desc, id desc"
     _mailing_enabled = True
@@ -391,6 +392,9 @@ class PosOrder(models.Model):
 
     _unique_uuid = models.Constraint('unique (uuid)', 'An order with this uuid already exists')
 
+    def ask_for_ticket_printing(self):
+        self.config_id._notify("TICKET_PRINTING_REQUESTED", self.ids)
+
     def get_preparation_change(self):
         self.ensure_one()
         return {
@@ -511,32 +515,34 @@ class PosOrder(models.Model):
     def _onchange_amount_all(self):
         self._compute_prices()
 
+    def _get_order_tax_totals(self):
+        self.ensure_one()
+        self.amount_paid = sum(payment.amount for payment in self.payment_ids)
+        self.amount_return = -sum((payment.amount < 0 and payment.amount) or 0 for payment in self.payment_ids)
+        base_lines = self.lines._prepare_tax_base_line_values()
+        self.env['account.tax']._add_tax_details_in_base_lines(base_lines, self.company_id)
+        self.env['account.tax']._round_base_lines_tax_details(base_lines, self.company_id)
+
+        cash_rounding = None
+        if (
+            self.config_id.cash_rounding
+            and not self.config_id.only_round_cash_method
+            and self.config_id.rounding_method
+        ):
+            cash_rounding = self.config_id.rounding_method
+
+        return self.env['account.tax']._get_tax_totals_summary(
+            base_lines=base_lines,
+            currency=self.currency_id,
+            company=self.company_id,
+            cash_rounding=cash_rounding,
+        )
+
     def _compute_prices(self):
-        AccountTax = self.env['account.tax']
         for order in self:
             if not order.currency_id:
                 raise UserError(_("You can't: create a pos order from the backend interface, or unset the pricelist, or create a pos.order in a python test with Form tool, or edit the form view in studio if no PoS order exist"))
-            order.amount_paid = sum(payment.amount for payment in order.payment_ids)
-            order.amount_return = -sum(payment.amount < 0 and payment.amount or 0 for payment in order.payment_ids)
-
-            base_lines = order.lines._prepare_tax_base_line_values()
-            AccountTax._add_tax_details_in_base_lines(base_lines, order.company_id)
-            AccountTax._round_base_lines_tax_details(base_lines, order.company_id)
-
-            cash_rounding = None
-            if (
-                order.config_id.cash_rounding
-                and not order.config_id.only_round_cash_method
-                and order.config_id.rounding_method
-            ):
-                cash_rounding = order.config_id.rounding_method
-
-            tax_totals = AccountTax._get_tax_totals_summary(
-                base_lines=base_lines,
-                currency=order.currency_id,
-                company=order.company_id,
-                cash_rounding=cash_rounding,
-            )
+            tax_totals = order._get_order_tax_totals()
             refund_factor = -1 if (order.amount_total < 0.0) else 1
             order.amount_tax = refund_factor * tax_totals['tax_amount_currency']
             order.amount_total = refund_factor * tax_totals['total_amount_currency']
@@ -1097,6 +1103,19 @@ class PosOrder(models.Model):
         for line in lines_to_reconcile.values():
             line.filtered(lambda l: not l.reconciled).reconcile()
 
+    @api.model
+    def get_example_order_data(self):
+        last_order = self.env['pos.order'].search([], order='id desc', limit=1)
+        return last_order.order_receipt_generate_data()
+
+    def action_pos_order_receipt(self):
+        self.ensure_one()
+        return {
+            "type": "ir.actions.act_url",
+            "url": f"/pos/receipt/{self.id}?company_id={self.company_id.id}",
+            "target": "new",
+        }
+
     def action_pos_order_invoice(self):
         self.ensure_one()
         if not (move := self.account_move):
@@ -1388,33 +1407,42 @@ class PosOrder(models.Model):
             'target': 'new'
         }
 
-    def action_send_receipt(self, email, ticket_image, basic_image):
+    def action_send_receipt(self, email):
         self.ensure_one()
         self.email = email
         mail_template_id = 'point_of_sale.email_template_pos_receipt'
         mail_template = self.env.ref(mail_template_id, raise_if_not_found=False)
+        ticket_image = self.order_receipt_generate_image()
+        basic_image = self.order_receipt_generate_image(True)
         if not mail_template:
             raise UserError(_("The mail template with xmlid %s has been deleted.", mail_template_id))
-        mail_template.send_mail(self.id, force_send=True, email_values={'email_to': email,
-                                                                        'attachment_ids': self._get_mail_attachments(self.name, ticket_image, basic_image)})
+        mail_template.send_mail(
+            self.id,
+            force_send=True,
+            email_values={
+                'email_to': email,
+                'attachment_ids': self._get_mail_attachments(self.name, ticket_image, basic_image)
+            })
 
     def _get_mail_attachments(self, name, ticket, basic_ticket):
         attachments = []
-        receipt = self.env['ir.attachment'].create({
-            'name': 'Receipt-' + name + '.jpg',
-            'type': 'binary',
-            'datas': ticket,
-            'res_model': 'pos.order',
-            'res_id': self.ids[0],
-            'mimetype': 'image/jpeg',
-        })
-        attachments += [(4, receipt.id)]
+
+        if ticket:
+            receipt = self.env['ir.attachment'].create({
+                'name': 'Receipt-' + name + '.jpg',
+                'type': 'binary',
+                'datas': base64.b64encode(ticket),
+                'res_model': 'pos.order',
+                'res_id': self.ids[0],
+                'mimetype': 'image/jpeg',
+            })
+            attachments += [(4, receipt.id)]
 
         if basic_ticket:
             basic_receipt = self.env['ir.attachment'].create({
                 'name': 'Receipt-' + name + '-1' + '.jpg',
                 'type': 'binary',
-                'datas': basic_ticket,
+                'datas': base64.b64encode(basic_ticket),
                 'res_model': 'pos.order',
                 'res_id': self.ids[0],
                 'mimetype': 'image/jpeg',
@@ -1667,12 +1695,13 @@ class PosOrderLine(models.Model):
             res = line._compute_amount_line_all()
             line.update(res)
 
-    def _compute_amount_line_all(self):
+    def _compute_amount_line_all(self, qty=None):
         self.ensure_one()
         fpos = self.order_id.fiscal_position_id
         tax_ids_after_fiscal_position = fpos.map_tax(self.tax_ids)
         price = self.price_unit * (1 - (self.discount or 0.0) / 100.0)
-        taxes = tax_ids_after_fiscal_position.compute_all(price, self.order_id.currency_id, self.qty, product=self.product_id, partner=self.order_id.partner_id)
+        line_qty = qty or self.qty
+        taxes = tax_ids_after_fiscal_position.compute_all(price, self.order_id.currency_id, line_qty, product=self.product_id, partner=self.order_id.partner_id)
         return {
             'price_subtotal_incl': taxes['total_included'],
             'price_subtotal': taxes['total_excluded'],

--- a/addons/point_of_sale/models/res_company.py
+++ b/addons/point_of_sale/models/res_company.py
@@ -38,7 +38,7 @@ class ResCompany(models.Model):
             'id', 'currency_id', 'email', 'website', 'company_registry', 'vat', 'name', 'phone', 'partner_id',
             'country_id', 'state_id', 'tax_calculation_rounding_method', 'nomenclature_id', 'point_of_sale_use_ticket_qr_code',
             'point_of_sale_ticket_unique_code', 'point_of_sale_ticket_portal_url_display_mode', 'street', 'city', 'zip',
-            'account_fiscal_country_id',
+            'account_fiscal_country_id', 'street2',
         ]
 
     @api.constrains('fiscalyear_lock_date', 'tax_lock_date', 'sale_lock_date', 'hard_lock_date')
